File: ParadigmaComp/15Questoes/golang/questao15.go
package main
//Exceções/Erros Personalizados
import (
	"errors"
	"fmt"
)

// Estrutura ContaBancaria
type ContaBancaria struct {
	Saldo float64
}

// Função para sacar dinheiro, retornando erro personalizado
func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > c.Saldo {
		return errors.New("saldo insuficiente")
	}
	c.Saldo -= valor
	return nil
}

func main() {
	conta := ContaBancaria{Saldo: 100}

	err := conta.Sacar(150)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("Saque realizado com sucesso")
	}
}

File: ParadigmaComp/15Questoes/python/questao15.py
#Exceções/Erros Personalizados
class SaldoInsuficienteException(Exception):
    def __init__(self, mensagem="Saldo insuficiente para a operação"):
        self.mensagem = mensagem
        super().__init__(self.mensagem)

class ContaBancaria:
    def __init__(self, saldo):
        self.saldo = saldo

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException()
        self.saldo -= valor

# Exemplo de uso
conta = ContaBancaria(100)
try:
    conta.sacar(150)
except SaldoInsuficienteException as e:
    print(e)
